class-tgm-plugin-activation.php

```diff
@@ -630,6 +632,73 @@ protected function do_plugin_install() {
 		}
 
 		/**
+		 * Adjust the plugin directory name if necessary.
+		 *
+		 * The final destination directory of a plugin is based on the subdirectory name found in the
+		 * (un)zipped source. In some cases - most notably GitHub repository plugin downloads -, this
+		 * subdirectory name is not the same as the expected slug and the plugin will not be recognized
+		 * as installed. This is fixed by adjusting the temporary unzipped source subdirectory name to
+		 * the expected plugin slug.
+		 *
+		 * @param string       $source        Path to upgrade/zipfilename.tmp/subdirectory/
+		 * @param string       $remote_source Path to upgrade/zipfilename.tmp
+		 * @param \WP_Upgrader $upgrader      Instance of the upgrader which installs the plugin
+		 *
+		 * @return string $source
+		 */
+		public function maybe_adjust_source_dir( $source, $remote_source, $upgrader ) {
+
+			if ( ! $this->is_tgmpa_page() ) {
+				return $source;
+			}
+
+			// Check for single file plugins
+			$source_files = array_keys( $GLOBALS['wp_filesystem']->dirlist( $remote_source ) );
+			if ( 1 === count( $source_files ) && false === $wp_filesystem->is_dir( $source ) ) {
+
+				return $source;
+			}
+
+			// Multi-file plugin, let's see if the directory is correctly named
+			$desired_slug = '';
+
+			// Figure out what the slug is supposed to be
+			if ( false === $upgrader->bulk ) {
+				$desired_slug = $upgrader->skin->options['plugin']['slug'];
+			}
+			else {
+				// Bulk installer contains less info, so fall back on the info registered here.
+				foreach ( $this->plugins as $plugin ) {
+					if ( $plugin['name'] === $upgrader->skin->plugin_names[ $upgrader->skin->i ] ) {
+						$desired_slug = $plugin['slug'];
+						break;
+					}
+				}
+			}
+
+			if ( '' !== $desired_slug ) {
+				$subdir_name = substr( str_replace( $remote_source . '/', '', $source ), 0, -1 );
+
+				if ( ! empty( $subdir_name ) && $subdir_name !== $desired_slug ) {
+					$from = substr( $source, 0, -1 ); // remove end slash
+					$to   = trailingslashit( $remote_source ) . $desired_slug;
+
+					if ( true === $GLOBALS['wp_filesystem']->move( $from, $to ) ) {
+						return trailingslashit( $to );
+					}
+					else {
+						return new WP_Error( 'rename_failed', __( 'The remote plugin package is does not contain a folder with the desired slug and renaming did not work. Please contact the plugin provider and ask them to package their plugin according to the WordPress guidelines.', 'tgmpa' ), array( 'found' => $subdir_name, 'expected' => $desired_slug ) );
```

Comment:
Should this string (and the one below) be moved with the rest of the strings and then simply referenced here?